### PR TITLE
Documentation, particularly for FSM Module

### PR DESCRIPTION
**Pull request by hannahhoward**
# Goals

Provide basic module documentation since that is a source of confusion
Provide guidance on using FSM module beyond examples in test

# Implementation

- Add a detailed README for FSM module cause I know it best
- Add a simple README for the generic state machine module since it's in use in `go-storage-miner`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,6 +2,71 @@
...
+
+A state machine is defined in terms of a Planner.
+
+It has the following signature:
+
+```golang
+type Planner func(events []Event, user interface{}) (interface{}, uint64, error)
+```
+
+A planner receives a series of events and a point to the current state (represented by user)
```
Comment (ingar) [inline README.md]: "pointer"?

File: README.md [hunk 2]
```diff
@@ -2,6 +2,71 @@
...
+- Take further actions based on new state, possibly generating more events
+
+## Usage
+
+A state machine is defined in terms of a Planner.
+
+It has the following signature:
+
+```golang
+type Planner func(events []Event, user interface{}) (interface{}, uint64, error)
```
Comment (laser) [inline README.md]: Suggestion: Give names to the return parameters.

File: README.md (hunk 1 above)
Comment (laser) [inline README.md]: > and a point to the current state

Suggestion: "and a pointer to the current state"

File: README.md [hunk 3]
```diff
@@ -2,6 +2,71 @@
...
+
+var stateMachines = statemachine.New(ds, stateHandler, StateType{})
+```
+
+This creates a machine for the given data store that will process data of type `StateType` with the given stateHandler.
+
+You can now dispatch events to a state machine with:
+
+```golang
+var id interface{} // some identifier of a deal
```
Comment (ingar) [inline README.md]: Should be identifier of a "tracked state" or something like that.

File: fsm/README.md [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+[Wikipedia - Finite State Machines](https://en.wikipedia.org/wiki/Finite-state_machine)
+[Wikipedia - UML State Machine](https://en.wikipedia.org/wiki/UML_state_machine)
+
+## Description
+
+This library provides a way to model a Filecoin process that operates on a specific data structure as a series of state transitions. It is used in the storage and retrieval markets in [go-fil-markets repository](https://github.com/filecoin-project/go-fil-markets) to model the lifecycle of a storage or retrieval deal. It may be used for other parts of Filecoin in the future.
+
+A state machine is defined in terms of
+
+- StateType -- the type of data structure we are tracking (should be a struct)
```
Comment (laser) [inline fsm/README.md]: Note: Some things in this list, e.g. Events, are backtick'ed - and others (like StateKeyField) are not.

File: fsm/README.md [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+	FundsReceived   abi.TokenAmount
+  // an informational message to augment the status of the deal
+  Message         string
+}
+```
+
+Let's pretend our ideal deal flow looks like:
+
+```
+Receive new deal proposal -> Validate proposal-> For All Data Requested, Send a chunk, then request payment, then wait for payment before sending more -> Complete deal
```
Comment (laser) [inline fsm/README.md]: Suggestion:

Change

```
Validate proposal->
```

to

```
Validate proposal ->
```

File: fsm/README.md [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+- `Environment` - This is a single interface that is used to access external dependencies. It is available to a `StateEntryFunc`
+- `Notifier` - A function that gets called on each successfully applied state transition, before any state entry func is called. This is useful for providing external notifications about state updates. It is called with the name of the event applied and the current state of the data structure.
+
+## Usage
+
+Let's consider a hypothetical deal we want to track in Filecoin. Each deal will have a series of states it can be in, and various things that can happen to change its state. We will track multiple deals at the same time. In this example, we will model the actions of the receiving party (i.e. the person who accepted the deal and responding)
```
Comment (ingar) [inline fsm/README.md]: the person who accepted the deal and **is** responding

File: fsm/README.md [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+
+You can imagine at each state in this happy path, things could go wrong. A deal proposal might need to be rejected for not meeting criteria. A network message could fail to send. A client could send only a partial payment, or one that fails to process. We might had an error reading our own data.
+
+You can start to assemble different events that might happen in this process:
+
+```
+ReceivedNewDeal
+AcceptedDeal
+RejectedDeal
+SentBlocks
```
Comment (ingar) [inline fsm/README.md]: Should be `SentData` to be consistent with below...

**Pull request merged**